### PR TITLE
refactored onKeyDown to be easier to read, and added test

### PR DESCRIPTION
**Pull request by brandonrainey**
#918 

Changed the onKeyDown function in search-display-controller to be easier to read. Also created a new test file to test preserved behavior between the current and new function. 

The test passes, but I'm not sure if the test is up to par or not, but I tried to replicate as well as I could figure out how to.

### CODE REVIEW

**Comment by Kuuuube** [comment]
Add this to the tests:
```
ctrl + a
alt + a
meta + a
backspace
ctrl + backspace
arrow key down
```

**Review by jamesmaa — COMMENTED** [review]
Thanks for doing this 😄 

My biggest concern stems from replicating the `search-display-controller` logic in the tests because even though they verify the correctness of the current yomitan, if the logic of those if statement changes in the future, we would need to update the corresponding tests, which I don't expect future developers to remember to do.

File: test/search-display-controller.test.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    ) {
+        queryInput.focus({preventScroll: true});
+        queryInputCount += 1;
+    }
+}
+
+/**
+ * @param {KeyboardEvent} e
+ */
+function onKeyDownNew(e) {
```
[the remark this answers is not in the document]
Comment (brandonrainey) [reply test/search-display-controller.test.js:74]: I originally tried to make an instance of SearchDisplayController in the test, but I have to import every dependency and its dependencies and so on, which seems like a lot. I've also tried some other ways but nothing has seemed to work to grab the function.

[the remark this answers is not in the document]
Comment (jamesmaa) [reply test/search-display-controller.test.js:74]: I'm fine ignoring the eslint error

File: test/search-display-controller.test.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        new KeyboardEvent('keydown', {key: ' ', ctrlKey: false, metaKey: false, altKey: false}),
+        new KeyboardEvent('keydown', {key: 'a', ctrlKey: true, metaKey: false, altKey: false}),
+        new KeyboardEvent('keydown', {key: 'a', ctrlKey: false, metaKey: true, altKey: false}),
+        new KeyboardEvent('keydown', {key: 'a', ctrlKey: false, metaKey: false, altKey: true}),
+        new KeyboardEvent('keydown', {key: 'Backspace'}),
+        new KeyboardEvent('keydown', {key: 'Backspace', ctrlKey: true, metaKey: false, altKey: false}),
+        new KeyboardEvent('keydown', {key: 'ArrowDown'})
+    ];
+
+    test('should test that onKeyDownCurrent function focuses input', () => {
```
Comment (jamesmaa) [inline test/search-display-controller.test.js:104]: btw I feel like we should include one or two tests that test the negative condition.

For example, if `onKeyDown`'s logic is 
```js
function onKeyDown(e){
    if (true) {
        queryInput.focus({preventScroll: true});
    }
}
```

Then these tests would pass even though it's wrong

File: test/search-display-controller.test.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+const queryInput = querySelectorNotNull(document, '#search-textbox');
+
+const focusSpy = vi.spyOn(queryInput, 'focus');
+
+describe('Keyboard Event Handling', () => {
+    afterAll(() => teardown(global));
+
+    const validKeypressEvents = [
+        new KeyboardEvent('keydown', {key: 'a', ctrlKey: false, metaKey: false, altKey: false}),
+        new KeyboardEvent('keydown', {key: 'Backspace'})
```
Comment (Kuuuube) [inline test/search-display-controller.test.js:68]:
```suggestion
        new KeyboardEvent('keydown', {key: 'Backspace'}),
        new KeyboardEvent('keydown', {key: 'Backspace', ctrlKey: true, metaKey: false, altKey: false})
```

File: ext/js/display/search-display-controller.js [hunk 4]
```diff
@@ -154,16 +154,14 @@ export class SearchDisplayController {
...
-            e.key !== ' '
-        ) {
+        const activeElement = document.activeElement;
+
+        const isInputField = this._isElementInput(activeElement);
+        const isAllowedKey = e.key.length === 1 || e.key === 'Backspace';
+        const isModifierKey = e.ctrlKey || e.metaKey || e.altKey;
+        const isSpaceKey = e.key === ' ';
+
+        if (!isInputField && !isModifierKey && isAllowedKey && !isSpaceKey) {
```
Comment (Kuuuube) [inline ext/js/display/search-display-controller.js:164]: Suggestion for how to handle the ctrl + backspace case. Feel free to disregard if you have a better idea.
```suggestion
        const isCtrlBackspace = e.ctrlKey && e.key === 'Backspace';
        const isModifierKey = e.ctrlKey || e.metaKey || e.altKey;
        const isSpaceKey = e.key === ' ';

        if (!isInputField && (!isModifierKey || isCtrlBackspace) && isAllowedKey && !isSpaceKey) {
```

**Review by Kuuuube — CHANGES_REQUESTED** [review]
It would also be good to include this test in the package.json. Maybe something like this:

```json
"test:onkeydown": "vitest run --config test/data/vitest.search-display-controller.config.json",
```

With `test/data/vitest.search-display-controller.config.json` looking like this:
```json
{
    "test": {
        "include": [
            "test/search-display-controller.test.js"
        ]
    }
}
```

Then `npm run test:search-display-controller` can be added to `"test"` command in package.json as well.

**Comment by Kuuuube** [comment]
Actually disregard the part about adding a specific script for this in the package. It's probably not important enough to need that. It runs with `test:js` anyways.

**Review by Kuuuube — APPROVED** [review]
Thank you for taking the time to add this!
